ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/BlockManager.java

```diff
@@ -163,11 +170,15 @@ private void importBlockIgnoringResult(final SignedBeaconBlock block) {
                         result.getFailureCause().orElse(null));
                     break;
                   default:
-                    LOG.trace(
-                        "Unable to import block for reason {}: {}",
-                        result.getFailureReason(),
-                        block);
-                    dropInvalidBlock(block);
+                    LOG.trace("Unable to import block for reason {}: {}", failureReason, block);
+                }
+
+                switch (failureReason.validity) {
+                  case OPTIMISTIC:
+                  case INVALID:
+                    applyFailureValidity(block, failureReason.validity);
+                    break;
+                  default:
```

Comment:
I feel like we might be overthinking this. If we know the parent is invalid (either by EL or CL rules) then we don't want to try importing the child block at all and should return either that the block failed import because it's invalid (if the parent is invalid by CL rules) or `FAILED_PAYLOAD_EXECUTION` (if the parent payload failed then the child payload fails too).  If the parent is either valid or optimistic, then we just return whatever result we get when we try to execute the child.

Reply:
I'd go one level deeper to where we currently handle invalid and known blocks.  So we currently have:

```
  private SafeFuture<BlockImportResult> doImportBlock(final SignedBeaconBlock block) {
    notifyReceivedBlockSubscribers(block);
    if (!shouldImportBlock(block)) {
      return SafeFuture.completedFuture(BlockImportResult.knownBlock(block));
    }
```

That `!shouldImportBlock` is currently checking if the block is already imported or already known to be invalid. We actually want to handle three separate cases:

 1. Block is already imported, return BlockValidationResult.knownBlock as we do now
 2. Block is already known to be invalid, return the BlockValidationResult we got from it last time
 3. Parent block is already known to be invalid, return the BlockValidationResult we got from the parent (same error reason anyway - need to be careful we don't wind up with an error message that makes it look like we were trying to import the parent again). 

 If we split that into two separate cases we could then have essentially the check you're suggesting, so roughly:

```
  private SafeFuture<BlockImportResult> doImportBlock(final SignedBeaconBlock block) {
    if (isInvalid(block.getRoot)) {
      return SafeFuture.completedFuture(<previousImportResult>);
    }
    if (isInvalid(block.getParentRoot())) {
      return SafeFuture.completedFuture(<blockImportResultFromParent>);
    }
    notifyReceivedBlockSubscribers(block);
    if (isKnownBlock(block)) {
      return SafeFuture.completedFuture(BlockImportResult.knownBlock(block));
    }
```